macaroonbakery/authorizer.py

```diff
@@ -0,0 +1,115 @@
+# Copyright 2017 Canonical Ltd.
+# Licensed under the LGPLv3, see LICENCE file for details.
+import abc
+
+# EVERYONE is recognized by ACLAuthorizer as the name of a
+# group that has everyone in it.
+EVERYONE = "everyone"
+
+
+class AuthorizerError(Exception):
+    """ AuthorizerError is the exception raised when when the authorization
+    cannot be.
+    """
+    pass
+
+
+class Authorizer(object):
+    ''' Used to check whether a given user is allowed to perform a set of
+    operations.
+    '''
+    __metaclass__ = abc.ABCMeta
+
+    @abc.abstractmethod
+    def authorize(self, id, ops):
+        ''' Checks whether the given identity (which will be None when there is
+        no authenticated user) is allowed to perform the given operations.
+        It should raise an exception only when the authorization cannot be
+        determined, not when the user has been denied access.
+
+        On success, each element of allowed holds whether the respective element
+        of ops has been allowed, and caveats holds any additional third party
+        caveats that apply.
+        If allowed is shorter then ops, the additional elements are assumed to
+        be False.
+        '''
+        raise NotImplementedError('users must define authorize to use this base'
+                                  ' class')
+
+
+class AuthorizerFunc(Authorizer):
+    ''' Implements a simplified version of Authorizer that operates on a single
+    operation at a time.
+    '''
+    def __init__(self, f):
+        self._f = f
+
+    def authorize(self, id, ops):
+        '''Implements Authorizer.authorize by calling f with the given identity
+        for each operation.
+        '''
+        allowed = []
+        caveats = []
+        for op in ops:
+            ok, fcaveats = self._f(id, op)
+            allowed.append(ok)
+            if fcaveats is not None:
+                caveats.extend(fcaveats)
+        return allowed, caveats
+
+
+class ACLAuthorizer(Authorizer):
+    ''' ACLAuthorizer is an Authorizer implementation that will check access
+    control list (ACL) membership of users. It uses get_acl to find out
+    the ACLs that apply to the requested operations and will authorize an
+    operation if an ACL contains the group "everyone" or if an ACLIdentity
```

Comment:
or if the identity is an instance of ACLIdentity and its allow method returns True for the ACL. ?

Reply:
done.